### PR TITLE
API for specifying custom gradients.

### PR DESCRIPTION
**Pull request by FedericoV**
As discussed here: https://github.com/blackjax-devs/blackjax/issues/201

I went ahead and tried adding the option to pass custom gradients, and added some basic tests that verify that things still work with custom gradients (in this case, I just used forward mode gradients).

Feedback on API very welcome - this was the simplest design I could think of.  I haven't finished changing the code everywhere, but wanted to hear comments on API before going further.

### CODE REVIEW

File: blackjax/mcmc/hmc.py [hunk 1]
```diff
@@ -63,11 +63,16 @@ class HMCInfo(NamedTuple):
...
 
 
-def init(position: PyTree, logprob_fn: Callable):
+def init(position: PyTree, logprob_fn: Callable, logprob_grad_fn: Optional[Callable] = None):
     def potential_fn(x):
         return -logprob_fn(x)
 
-    potential_energy, potential_energy_grad = jax.value_and_grad(potential_fn)(position)
+    if logprob_grad_fn:
+        potential_energy_grad = -logprob_grad_fn(position)
```
Comment (junpenglao) [inline blackjax/mcmc/hmc.py:71]: comment to remind us why the negative sign here.

**Comment by FedericoV** [comment]
Added some more tests: I see a test failure that I believe is due to a numerical issue, but, another set of eyes is welcome.

File: tests/test_sampling.py [hunk 2]
```diff
@@ -244,15 +254,21 @@ def mcse_test(self, samples, true_param, p_val=0.01):
...
         )
         num_chains = 10
         initial_positions = jax.random.normal(pos_init_key, [num_chains, 2])
-        kernel = algorithm(logprob_fn, inverse_mass_matrix=true_scale, **parameters)
+        if custom_gradients:
+            logprob_grad_fn = jax.jacfwd(logprob_fn)
+            kernel = algorithm(logprob_fn, inverse_mass_matrix=true_scale, logprob_grad_fn=logprob_grad_fn,
+                **parameters)
+        else:
+            kernel = algorithm(logprob_fn, inverse_mass_matrix=true_scale, **parameters)
```
Comment (junpenglao) [inline tests/test_sampling.py:270]:
```suggestion
        if custom_gradients:
            logprob_grad_fn = jax.jacfwd(logprob_fn)
        else:
            logprob_grad_fn = None
        kernel = algorithm(logprob_fn, inverse_mass_matrix=true_scale, logprob_grad_fn=logprob_grad_fn,
                **parameters)
```

**Comment by rlouf** [comment]
I made a few changes (mostly naming and linting), LGTM otherwise. Is there anything else to change in the code?

**Comment by FedericoV** [comment]
There was a test that was failing when switching reverse mode gradients to forward mode gradients: do you think that's a cause for concern?  It just looked like a numerical tolerance issue to me but wanted to double check.

The only other issue is that there's a little bit of repeated code, but, it's probably not worth factoring it out yet.

**Comment by junpenglao** [comment]
@rlouf How do you make the test not skip? The test are not run right now
@FedericoV We should try to get a better understanding of why test fail for sure, maybe even enable the custom_grad test for a few more places.

**Comment by rlouf** [comment]
It's an issue with the github workflow that I need to fix. Tests are passing locally btw.

**Comment by FedericoV** [comment]
Let me know if you'd like to add any more changes.  I'm fine with merging it whenever.

**Comment by rlouf** [comment]
Thank you for contributing!

**Comment by FedericoV** [comment]
Awesome, glad to help!
